src/cursor.rs

```diff
@@ -251,3 +253,121 @@ impl<'a> Iterator for TailingCursor<'a> {
         }
     }
 }
+
+type DocArray = VecDeque<Document>;
+type CursorId = i64;
+
+pub struct BatchCursor<'a> {
+    cursor:     Cursor<'a>,
+    db:         &'a Database<'a>,
+    coll_name:  String,
+    cursor_id:  Option<CursorId>,
+    documents:  Option<DocArray>
+
+}
+
+impl<'a> BatchCursor<'a> {
+    pub fn new(
+        cursor: Cursor<'a>,
+        db: &'a Database<'a>,
+        coll_name: String
+    ) -> BatchCursor<'a> {
+        BatchCursor {
+            cursor,
+            db,
+            coll_name,
+            cursor_id: None,
+            documents: None
+        }
+    }
+
+    fn get_cursor_next(&mut self) -> Option<Result<Document>> {
+        let item_opt = self.cursor.next();
+        if let Some(item_res) = item_opt {
+            if let Ok(item) = item_res {
+                let docs_ret = batch_to_array(item);
+                if let Ok(docs) = docs_ret {
+                    self.documents = docs.0;
+                    if docs.1.is_some() {self.cursor_id = docs.1}
+                    let res = self.get_next_doc();
+                    if res.is_some() { return res; }
+                } else {
+                    return Some(Err(docs_ret.err().unwrap()));
+                }
+            }
+        }
+        None
+    }
+
+    fn get_next_doc(&mut self) -> Option<Result<Document>> {
```

Comment:
Why two function for getting the next thing? Why would I use the one or the other? Can we document that?

Reply:
I added it to the documentation.  these are both internal helper functions.  one is for populating the internal documents buffer.  the other is for reading from that buffer.

Reply:
Sorry, missed that this one is private. Ignore what I said.